sympy/polys/sqfreetools.py

```diff
@@ -118,59 +184,203 @@ def dup_sqf_norm(f, K):
     return s, f, r
 
 
+def _dmp_sqf_norm_shifts(f, u, K):
+    """Generate a sequence of candidate shifts for dmp_sqf_norm."""
+    #
+    # We want to find a minimal shift if possible because shifting high degree
+    # variables can be expensive e.g. x**10 -> (x + 1)**10. We try a few easy
+    # cases first before the final infinite loop that is guaranteed to give
+    # only finitely many bad shifts (see Trager76 for proof of this in the
```

Comment:
Probably not worth it, but one could imagine more elaborate strategies for exploring the n-dimensional lattice, where we'd visit more of the points that avoid shifting high-degree variables, before moving on.

Like for `f = x^10 + y^2 + z` the existing strategy tries

```
[0, 0, 1]
[0, 1, 0]
[1, 0, 0]
```

but maybe it would be better to try

```
[0, 1, 1]
```

before

```
[1, 0, 0]
```

Anyway, again, probably not worth it. Interesting though.

Reply:
I have considered this as well. I'm not sure what statistics are for how often different versions actually work e.g. for the example
```python
(x**10 + y**2 + z).as_poly(domain=QQ.algebraic_field(sqrt(2))).sqf_norm()
```
What actually happens is that we try `[0,0,0]` and then `[0,0,1]` and then it terminates. I think that for randomly generated polynomials it is quite likely that either the first or second iteration gives us the square-free norm.

I have added a test case that gets us to the final step of choosing shifts but it is contrived by multiplying together independent polynomials e.g. this case gets there:
```python
((x**10 + 1)*(y**2 + 1)*z).as_poly(domain=QQ.algebraic_field(sqrt(2))).sqf_norm()
```
In this case the shifts tried are
```python
[0, 0, 0]
[0, 0, 1]
[0, 1, 0]
[1, 0, 0]
[1, 1, 1]
```
Here the fact that the factors each contain only one variable makes it much more likely that we need to shift all variables to get a square free norm. The more cases we try before `[1, 1, 1]` the more time we waste. I assume that this is the more likely cause of cases where we need more than just a shift of one variable.

Reply:
Okay, so moving on sooner could actually be a good thing. That's interesting. Well then, what you have seems good.